### PR TITLE
feat: add Histogram metrics to metrics crate

### PR DESCRIPTION
**Pull request by e-dard**
This PR adds a Histogram metric type to the metrics crate, resulting in Prometheus histograms (which are not samples).

The histogram metric type is good for tracking latencies and other per-request or per-operation side-effects that are best handled as a distribution. 

*TODO* - I can't figure out how to set the per-metric bucket boundaries so I suggest that gets added in a future PR. For now I have left some code commented out and all Histogram metrics will use the default buckets, which should be OK at least for request latencies.

### CODE REVIEW

File: metrics/src/metrics.rs [hunk 1]
```diff
@@ -273,3 +273,73 @@ impl Counter {
...
+    }
+
+    /// Add a new observation to the histogram
+    pub fn observe(&self, observation: f64) {
+        self.observe_with_labels(observation, &[]);
+    }
+
+    /// A helper method for observing latencies. Returns a new timing instrument
+    /// which will handle submitting an observation containing a duration.
+    pub fn timer(&self) -> HistogramTimer<'_> {
```
Comment (domodwyer) [inline metrics/src/metrics.rs:319]: Nice 👍

**Review by domodwyer — APPROVED** [review]
🕐

**Review by jacobmarble — APPROVED** [review]

File: metrics/src/lib.rs [hunk 2]
```diff
@@ -236,6 +236,85 @@ impl Domain {
...
+        }
+    }
+
+    /// Initialise the Histogram metric ready for observations.
+    pub fn init(self) -> metrics::Histogram {
+        let labels = self.labels.unwrap_or_default();
+        let boundaries = self.boundaries.unwrap_or_default();
+
+        if !boundaries.is_empty() {
+            todo!("histogram boundaries not yet configurable.");
```
Comment (mkmik) [inline metrics/src/lib.rs:313]: I don't understand;

which boundaries will be used?

**Review by mkmik — APPROVED** [review]
I don't fully understand how the metrics interact with the prometheus exporter, but as far as I can tell this is ok

File: metrics/src/lib.rs (hunk 2 above)
Comment (mkmik) [reply metrics/src/lib.rs:313]: I assume the default boundaries we set up in our export pipeline will be used

```
        PrometheusExporter::new(
            registry,
            controller,
            default_summary_quantiles,
            default_histogram_boundaries,
        )
```

Comment (e-dard) [reply metrics/src/lib.rs:313]: Yes they will. I need to figure out how to be able to over-ride them.

**Pull request merged**